### PR TITLE
Update message header structure

### LINKED ISSUE #1692: Update MessageHeader format to match Weave Message Format baseline
#### Problem
Given the[ Weave Message Format specification](https://github.com/openweave/openweave-core/blob/master/doc/specs/weave-message-format.pdf) as baseline, the CHIP [Message Header](https://github.com/project-chip/connectedhomeip/blob/master/src/transport/MessageHeader.cpp#L30) has reordered some fields and is missing the Exchange fields (such as Profile ID, etc.)

 #### Proposed Solution
Recast the CHIP Message Header fields to align with the baseline specification.  Also update all tests and examples as appropriate to behave as before but with the updated line format.

[BEFORE]
```
/**********************************************
 * Header format (little endian):
 *
 *  16 bit: | VERSION: 4 bit | FLAGS: 4 bit | RESERVED: 8 bit |
 *  16 bit: | Secure message type                             |
 *  32 bit: | MESSAGE_ID                                      |
 *  32 bit: | Secure Session ID                               |
 *  64 bit: | Message Authentication Tag                      |
 *  64 bit: | SOURCE_NODE_ID (iff source node flag is set)    |
 *  64 bit: | DEST_NODE_ID (iff destination node flag is set) |
 *
 **********************************************/
 ```
 
[AFTER -- Baseline compatible]
```
/**********************************************
 * Header format (little endian):
 *
 *  16 bit: | VERSION: 4 bit | FLAGS: 4 bit | RESERVED: 8 bit |
 *  32 bit: | MESSAGE_ID                                      |
 *  64 bit: | SOURCE_NODE_ID (iff source node flag is set)    |
 *  64 bit: | DEST_NODE_ID (iff destination node flag is set) |
 *   8 bit: | Exchange Header                                 | +add
 *   8 bit: | Exchange Message Type                           | (formally 16-bit Secure message type)
 *  16 bit: | Exchange ID                                     | (formally 32-bit Secure Session ID)
 *  32 bit: | Message Profile ID                              | +add
 *  64 bit: | Message Authentication Tag                      |
 **********************************************/
 ```

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
The message header structure wire format is missing some fields. And, some fields are at incorrect offsets.

 #### Summary of Changes
Update the message header structure to follow agreed upon format. Also, a part of structure is expected to be encrypted. Added support for the encrypted header fields.

fixes #1692

### CODE REVIEW

**Review by turon — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by kedars — APPROVED** [review]

File: src/transport/MessageHeader.cpp [hunk 1]
```diff
@@ -30,13 +30,21 @@
...
+ * -------- Unencrypted header -----------------------------------------------------
+ *  16 bit: | VERSION: 4 bit | FLAGS: 4 bit | ENCRYPTTYPE: 4 bit | RESERVED: 4 bit |
+ *  32 bit: | MESSAGE_ID                                                           |
+ *  64 bit: | SOURCE_NODE_ID (iff source node flag is set)                         |
+ *  64 bit: | DEST_NODE_ID (iff destination node flag is set)                      |
+ *  16 bit: | Encryption Key ID                                                    |
+ *  16 bit: | Payload Length                                                       |
+ * -------- Encrypted header -------------------------------------------------------
+ *  8 bit:  | Exchange Header                                                      |
+ *  8 bit:  | Exchange Message Type                                                |
```
Comment (gerickson) [inline src/transport/MessageHeader.cpp:42]: Exchange Message type seems to be a new field previously undefined by the Weave Message Layer. What is this and how is it to be used?

**Review by gerickson — CHANGES_REQUESTED** [review]
Code looks good; one open question.

File: src/transport/MessageHeader.cpp (hunk 1 above)
Comment (pan-apple) [reply src/transport/MessageHeader.cpp:42]: @gerickson this issue (https://github.com/project-chip/connectedhomeip/issues/1692) defines the field.

Comment (pan-apple) [reply src/transport/MessageHeader.cpp]: Changed it to MessageType.

Comment (gerickson) [reply src/transport/MessageHeader.cpp]: It sounds like from talking to @turon, this was a simple typo. Thanks for the correction and clarification.

**Comment by gerickson** [comment]
The final open issue then appears to be that if we are aligning to the baseline that this is missing a 32-bit profile identifier field with its constituent 16-bit sub-fields.

**Comment by woody-apple** [comment]
@jelderton @andy31415 @saurabhst ?

**Review by saurabhst — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Comment by yufengwangca** [comment]
> The final open issue then appears to be that if we are aligning to the baseline that this is missing a 32-bit profile identifier field with its constituent 16-bit sub-fields.

Suggest to use 'protocol identifier' instead of 'profile identifier' to avoid confusion since profile has different meanings in the context of Weave  and ZCL. In weave, profile represents the concept of protocol, and in ZCL, it represents the type of device.

**Comment by gerickson** [comment]
> > The final open issue then appears to be that if we are aligning to the baseline that this is missing a 32-bit profile identifier field with its constituent 16-bit sub-fields.
> 
> Suggest to use 'protocol identifier' instead of 'profile identifier' to avoid confusion since profile has different meanings in the context of Weave and ZCL. In weave, profile represents the concept of protocol, and in ZCL, it represents the type of device.

For that matter, perhaps we just call it an `unsolicited dispatch identifier`.

**Comment by yufengwangca** [comment]
> > > The final open issue then appears to be that if we are aligning to the baseline that this is missing a 32-bit profile identifier field with its constituent 16-bit sub-fields.
> > 
> > 
> > Suggest to use 'protocol identifier' instead of 'profile identifier' to avoid confusion since profile has different meanings in the context of Weave and ZCL. In weave, profile represents the concept of protocol, and in ZCL, it represents the type of device.
> 
> For that matter, perhaps we just call it an `unsolicited dispatch identifier`.

Sounds good, `unsolicited dispatch identifier` is a 32 bits integer to identify to which protocol the message need to be dispatched.

**Comment by gerickson** [comment]
> > > > The final open issue then appears to be that if we are aligning to the baseline that this is missing a 32-bit profile identifier field with its constituent 16-bit sub-fields.
> > > 
> > > 
> > > Suggest to use 'protocol identifier' instead of 'profile identifier' to avoid confusion since profile has different meanings in the context of Weave and ZCL. In weave, profile represents the concept of protocol, and in ZCL, it represents the type of device.
> > 
> > 
> > For that matter, perhaps we just call it an `unsolicited dispatch identifier`.
> 
> Sounds good, `unsolicited dispatch identifier` is a 32 bits integer to identify to which protocol the message need to be dispatched.

Which still has two sub-fields:

* 16-bit vendor identifier (administered by the Alliance)
* 16-bit vendor unsolicited dispatch identifier (administered by the vendor)

**Comment by woody-apple** [comment]
@gerickson @yufengwangca - given this keeps coming up, would you be ok with @pan-apple filing a new issue for "integrating Weave Profile ID" and tackling that separately - so he can make progress on SPAKE2+ pairing? I don't believe the lack of Weave ProfileID is blocking development at this juncture? My understanding is it has no impact to Cluster or Security progress - which are the critical paths right now. Is that accurate?

**Comment by pan-apple** [comment]
@gerickson, @yufengwangca  I've updated the header based on agreements in Secure Channel F2F meeting.

**Review by gerickson — APPROVED** [review]
Thanks for adding the additional baseline 32-bits for unsolicited protocol dispatch.

I suspect we'll need to modify the implementation a bit to handle the 0x0000 (CHIP "Common") vendor ID case being the same as no vendor ID; however, that's an implementation tweak that can be added later.

I also suspect we'll want a getter that can grab both fields as a convolved 32-bit value. Again, minor, later implementation tweak.

**Review by andy31415 — APPROVED** [review]

**Comment by woody-apple** [comment]
Thanks all!

**Pull request merged**